olp-cpp-sdk-core/src/client/HRN.cpp

```diff
@@ -19,84 +19,101 @@
 
 #include "olp/core/client/HRN.h"
 
-#include <cassert>
 #include <sstream>
 
+#include <olp/core/logging/Log.h>
+#include <olp/core/porting/make_unique.h>
 #include "Tokenizer.h"
 
+namespace {
+constexpr auto kLogTag = "HRN";
+static const std::string kDataTag = "data";
+static const std::string kPipelineTag = "pipeline";
+static const std::string kSchemaTag = "schema";
+static const std::string kHRNTag = "hrn:";
+static constexpr char kSeparator = ':';
+}  // namespace
+
 namespace olp {
 namespace client {
 
 std::string HRN::ToString() const {
   std::ostringstream ret;
-  std::string generic_part = ":" + region + ":" + account + ":";
-  ret << "hrn:" << partition << ":";
+  auto generic_part = kSeparator + region + kSeparator + account + kSeparator;
+  ret << kHRNTag << partition << kSeparator;
+
   switch (service) {
     case ServiceType::Data: {
-      ret << "data" << generic_part << catalogId;
+      ret << kDataTag << generic_part << catalogId;
       if (!layerId.empty()) {
-        ret << ":" << layerId;
+        ret << kSeparator << layerId;
       }
       break;
     }
-    case ServiceType::Schema:
-      ret << "schema" << generic_part << groupId << ":" << schemaName << ":"
-          << version;
+    case ServiceType::Schema: {
+      ret << kSchemaTag << generic_part << groupId << kSeparator << schemaName
+          << kSeparator << version;
       break;
-    case ServiceType::Pipeline:
-      ret << "pipeline" << generic_part << pipelineId;
+    }
+    case ServiceType::Pipeline: {
+      ret << kPipelineTag << generic_part << pipelineId;
       break;
-    default:
+    }
+    default: {
       ret << generic_part;
       break;
+    }
   }
+
   return ret.str();
 }
 
 std::string HRN::ToCatalogHRNString() const {
   if (service != ServiceType::Data) {
-    return std::string();
+    OLP_SDK_LOG_WARNING_F(kLogTag, "ToCatalogHRNString: ServiceType != Data");
+    return {};
   }
 
-  return "hrn:" + partition + ":data:" + region + ":" + account + ":" +
-         catalogId;
+  return kHRNTag + partition + ":data:" + region + kSeparator + account +
+         kSeparator + catalogId;
 }
 
 HRN::HRN(const std::string& input) {
-  Tokenizer tokenizer(input, ':');
+  Tokenizer tokenizer(input, kSeparator);
 
-  // hrn must start with "hrn:"
+  // Must start with "hrn:"
   if (!tokenizer.HasNext()) {
     return;
   }
 
-  std::string protocol = tokenizer.Next();
-
+  auto protocol = tokenizer.Next();
```

Comment:
```suggestion
  const auto protocol = tokenizer.Next();
```
Prefer immutable objects by default